storage/src/key_value/agent/mod.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    any::Any,
+    collections::{HashMap, VecDeque},
+};
+
+use snarkvm_dpc::{
+    testnet1::{instantiated::Components, Record},
+    BlockHeaderHash,
+    StorageError,
+};
+use snarkvm_utilities::{has_duplicates, FromBytes, ToBytes};
+use tokio::sync::mpsc;
+use tracing::log::trace;
+
+use crate::{
+    key_value::*,
+    BlockStatus,
+    CanonData,
+    Digest,
+    ForkDescription,
+    ForkPath,
+    KeyValueStorage,
+    SerialBlock,
+    SerialBlockHeader,
+    SerialTransaction,
+    TransactionLocation,
+    VMRecord,
+};
+
+mod block;
+mod block_commit;
+
+pub(super) struct Agent<S: KeyValueStorage + 'static> {
+    inner: S,
+}
+
+impl<S: KeyValueStorage + 'static> Agent<S> {
+    pub(super) fn new(inner: S) -> Self {
+        Agent { inner }
+    }
+
+    fn read_u32(value: &[u8]) -> Result<u32> {
+        if value.len() != 4 {
+            return Err(anyhow!("invalid length for u32: {}"));
+        }
+        let mut out = [0u8; 4];
+        out.copy_from_slice(&value[..]);
+        Ok(u32::from_le_bytes(out))
+    }
+
+    fn read_meta_u32(&mut self, name: &str, default: Option<u32>) -> Result<u32> {
+        let tip = self.inner.get(KeyValueColumn::Meta, name.as_bytes())?;
+        match tip {
+            Some(value) => Self::read_u32(&value[..]),
+            None => match default {
+                None => Err(anyhow!("missing meta value for {}", name)),
+                Some(default) => {
+                    self.inner
+                        .store(KeyValueColumn::Meta, name.as_bytes(), &default.to_le_bytes()[..])?;
+                    Ok(default)
+                }
+            },
+        }
+    }
+
+    fn write_meta_u32(&mut self, name: &str, value: u32) -> Result<()> {
+        self.inner
+            .store(KeyValueColumn::Meta, name.as_bytes(), &value.to_le_bytes()[..])?;
+        Ok(())
+    }
+
+    fn get_record_commitments(&mut self, limit: Option<usize>) -> Result<Vec<Digest>> {
+        let mut records = self.inner.get_column_keys(KeyValueColumn::Records)?;
+        if let Some(limit) = limit {
+            records.truncate(limit);
```

Comment:
as with a similar truncation earlier on, are these values sorted, resulting in a meaningful result of truncating to a given limit?

Reply:
mirrors existing logic, but i generally agree. this feature of storing records is a hack for miners and an ideal order isn't clear. feel free to open a github issue